virtual_rainforest/models/abiotic_simple/simple_regression.py

```diff
@@ -0,0 +1,714 @@
+        dim="layers",
+    )
+    atmospheric_co2 = atmospheric_co2_1.assign_coords(
+        {
+            "layers": np.arange(0, len(layer_roles)),
+            "layer_roles": ("layers", layer_roles[0 : len(layer_roles)]),  # noqa
+            "cell_id": data.grid.cell_id,
+        },
+    )
+    output.append(atmospheric_co2)
+
+    # precipitation at the surface is reduced as a function of leaf area index
+    precipitation_surface = data["precipitation"].isel(time=time_index) * (
+        1 - water_interception_factor * data["leaf_area_index"].sum(dim="layers")
+    )
+
+    # Soil moisture profile, []
+    soil_moisture, surface_run_off = calculate_soil_moisture(
+        layer_roles=layer_roles,
+        precipitation_surface=precipitation_surface,
+        current_soil_moisture=data["soil_moisture"],
+        soil_moisture_capacity=soil_moisture_capacity,
+    )
+
+    output.append(soil_moisture)
+    output.append(surface_run_off)
+    return output
+
+
+# supporting functions
+def lai_regression(
+    reference_data: DataArray,
+    leaf_area_index: DataArray,
+    gradient: float,
+) -> DataArray:
+    """Calculate microclimatic variable at 1.5 m as function of leaf area index.
+
+    Args:
+        reference_data: input variable at reference height
+        leaf_area_index: leaf area index, [m m-1]
+        gradient: gradient of regression from :cite:t:`hardwick_relationship_2015`
+
+    Returns:
+        microclimatic variable at 1.5 m
+    """
+
+    return DataArray(
+        leaf_area_index.sum(dim="layers") * gradient + reference_data, dims="cell_id"
+    )
+
+
+def logarithmic(x: DataArray, gradient: float, intercept: float) -> DataArray:
+    """Logarithmic function.
+
+    Args:
+        x: x values
+        gradient: gradient
+        intercept: intercept
+
+    Returns:
+        y values
+    """
+    return DataArray(gradient * np.log(x) + intercept)
+
+
+def log_interpolation(
+    data: Data,
+    reference_data: DataArray,
+    layer_roles: List[str],
+    layer_heights: DataArray,
+    value_from_lai_regression: DataArray,
+) -> DataArray:
+    """Logarithmic interpolation of variables for vertical profile.
+
+    Args:
+        data: Data object
+        reference_data: input variable at reference height
+        layer_roles: list of layer roles (soil, surface, subcanopy, canopy, above)
+        layer_heights: vertical layer heights, [m]
+        value_from_lai_regression: variable value from linear regression with LAI
+
+    Returns:
+        vertical profile of provided variable
+    """
+
+    # Fit logarithmic function to interpolate between temperature top and 1.5m
+    x_values = np.array(
+        [
+            layer_heights.isel(layers=0),
+            np.repeat(1.5, len(data.grid.cell_id)),
+        ]
+    ).flatten()
+    y_values = np.array([reference_data, value_from_lai_regression]).flatten()
+
+    popt, pcov = curve_fit(logarithmic, x_values, y_values)
+    a, b = popt  # the function coefficients
```

Comment:
I don't fully understand this - we have the (e.g.) temp at the canopy top and then the predicted temp given LAI at 1.5m and we expect intermediate heights to follow an exponential model.  I think this is using `curve_fit` to estimate that exponential model as an average across all cells, where actually it should be a per cell relationship?

Reply:
You are correct, I made a mistake here. The exponential model should be fitted for each grid cell individually. How could I best modify this? By looping over grid cells?